### PR TITLE
Save multiple datasets into a single ISIS ORSO file

### LINKED ISSUE #37002: Add flood and calibration info to SaveISISReflectometryORSO docs
We added flood and calibration file metadata to the ISIS ORSO file output by `SaveISISReflectometryORSO` in PR #36993. In the algorithm documentation, we should add to the header metadata table under `measurement additional_files` to mention that this now includes flood and calibration files.

### PR DESCRIPTION
**Pull request by rbauststfc**
#### Summary of work

This PR extends the `SaveISISReflectometryORSO` algorithm so that it can take a list of workspace names (or names of workspace groups) and will save all the datasets into a single file. The change to the input workspace parameter name in this PR is a breaking change, but the algorithm has only recently been added and is unlikely to have been used in scripts, so it seems sufficient to flag this up in the release notes rather than version the algorithm.

This PR is the first part of fixing issue #36396 and it focuses on the changes required to the algorithm only. The ISIS Reflectometry GUI is not updated to support outputting multiple datasets in a single file yet, so it should continue working as before. Updating the GUI will be done in a separate PR as part two of this work.

This PR also includes documentation updates, including some documentation updates that I missed when I added calibration and flood file metadata previously (#37002).

Refs #36396.
Fixes #37002. 

#### Further detail of work

I've refactored the code a bit to introduce a new `ReflectometryDataset` object. The idea is that this allows us to retrieve all the workspace history algorithms that we need up-front and then pass that around to use in the rest of the code. This is to avoid needing to search for algorithms in the workspace history multiple times in order to find different bits of metadata. It also allows us to easily set flags like identifying a workspace as being a member of a WorkspaceGroup, for example.

Another refactor was to change some of the log levels from warning to debug, to avoid excessive warnings when we eventually use this in the GUI for processing large numbers of runs.

A couple of other decisions that are worth flagging up:

- I agreed with our scientists that we would not validate the combination of workspaces and workspace groups that they pass in. However, as explained in the algorithm documentation, if the automatically generated dataset names for the list of workspaces ends up containing duplicates then the algorithm will throw an error. We can develop the algorithm further to add relevant validation or resolve duplicate names if use cases arise, but under normal, expected use this shouldn't be an issue so it seemed worth keeping the implementation simple to start with.

- If additional columns are requested but cannot be calculated then we still include them as nan values. I've done this, rather than excluding the columns, because when loading into analysis software an eight column file enables functionality that is not available for a four column file. This is something I intend to find out more about from our scientists and will change the approach in the part two PR if necessary. Note that 8 column datasets are only relevant for data that hasn't been stitched, so the additional columns option only has an effect for non-stitched data.

### To test:

1) Go to `Interfaces -> Reflectometry -> ISIS Reflectometry`.
1) On the Runs tab, enter Investigation ID `1120015` and Cycle `11_3`. Click the Search button.
1) From the search results, click to highlight runs `11934`, `11935` and `11936` and click the Transfer button to transfer them into the batch table on the right hand side. Click the Process button to reduce the data.
1) Go to the Save ASCII tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO ASCII (*.ort)` from the drop-down.
1) From the List of Workspaces section on the left, select workspaces `IvsQ_11934_11935+11936` and  `IvsQ_binned_11934`.
1) Click the big Save button on the right hand side to save to file. There should be a successful call to `SaveISISReflectometryORSO` recorded in the messages pane. If you navigate to the save path that you set in the Save ASCII tab, you should see two separate `.ort` files with one dataset in each. This tests that the GUI is still working as before (i.e. outputting multiple datasets to separate files).
1) Now close/minimise the Reflectometry GUI but keep the reduced workspaces in your workspace list. Call algorithm `SaveISISReflectometryORSO` from the algorithm dialog (or from script, if preferred) and enter the two workspace names you selected from the GUI. Set column `IncludeAdditionalColumns` to `True`. Run the algorithm to save out a file. This time the file should contain two datasets. The first should be the individual angle dataset with the theta value given as the dataset name. There should be 8 columns for this first dataset: Qz (1/angstrom), R, sR, sQz, lambda (angstrom), slambda, incident theta (degrees), sincident theta. The second dataset should be called "Stitched" and have 4 populated columns: Qz (1/angstrom), R, sR, sQz. **Note that because of an issue in the orsopy library, the stitched dataset will still have 8 column headers even though only four will have values. This bug has been reported to orsopy and we're waiting for the fix to be released. We'll pull this in when it becomes available.**
1) To test with a workspace group, go back to the ISIS Reflectometry GUI, change the instrument to `POLREF` and enter run number `31832` and angle `0.5`. Reduce this run and then test as above but instead pass `IvsQ_binned_31832` (the workspace group name) to the algorithm. This should save a single file with two datasets. This time the dataset names should contain the individual workspace name and the angle. This is not stitched data, so both datasets should have 8 columns.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by jclarkeSTFC — CHANGES_REQUESTED** [review]
Works as described, I followed the test instructions, built the docs, and checked the output from the GUI. Just a couple of very minor things I noticed.

File: docs/source/algorithms/SaveISISReflectometryORSO-v1.rst [hunk 1]
```diff
@@ -33,14 +52,24 @@ The resolution (dQ/Q) is looked up from the workspace history as follows:
...
-If a resolution value cannot be found from the workspace history then a warning is logged and the file is saved without this column included.
+If a resolution value cannot be found from the workspace history then the file is saved without this column included.
+
+If parameter ``IncludeAdditionalColumns`` is set to ``True`` then the value of parameter ``WriteResolution`` is ignored and the algorithm will output the four columns described above for stitched datasets.
+For non-stitched datasets there will be the four columns described above plus an additional four columns as follows:
+
+- *lambda* - the wavelength values. If the original conversion to Q was performed using :ref:`algm-RefRoi` then the Qz column values are converted back to wavelength using: :math:`\lambda=\frac{4\pi}{Q}sin(\theta)`. If the original conversion was performed using :ref:`algm-ConvertUnits` then this algorithm is used to convert back to wavelength.
+- *error of lambda* - currently assumed to be 0.
+- *incident theta* - the value of theta used for the final conversion to Q.
+- *error of incident theta* - calculated as :math:`resolution (dQ/Q) * \theta`.
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply docs/source/algorithms/SaveISISReflectometryORSO-v1.rst:63]: Sounds good. I think I'm being over-anxious by including the dQ/Q here, as it's included earlier in the text to confirm that we're talking about dQ/Q when we refer to resolution, so perhaps I'll just take it out altogether from this equation?

[the remark this answers is not in the document]
Comment (jclarkeSTFC) [reply docs/source/algorithms/SaveISISReflectometryORSO-v1.rst]: Yes that sounds good

**Review by jclarkeSTFC — APPROVED** [review]

**Comment by rbauststfc** [comment]
@mantidproject/gatekeepers would it be possible for a gatekeeper to take a look at this one? Thanks

**Pull request merged**